plugin/etcd/plugin.go

```diff
@@ -0,0 +1,496 @@
+				Title:   "Client Private Key",
+				Example: "-----BEGIN RSA PRIVATE KEY-----\n(cert contents)\n(... etc ...)\n-----END RSA PRIVATE KEY-----",
+			},
+			plugin.Field{
+				Mode:    "target",
+				Name:    "ca_cert",
+				Type:    "pem-x509",
+				Help:    "CA certificate in pem format that issued the client cert and key.",
+				Title:   "CA Certificate",
+				Example: "-----BEGIN CERTIFICATE-----\n(cert contents)\n(... etc ...)\n-----END CERTIFICATE-----",
+			},
+			plugin.Field{
+				Mode:    "target",
+				Name:    "overwrite",
+				Type:    "bool",
+				Help:    "If this is enabled, only keys mentioned in the 'Prefix' field will be deleted. The values will be restored using the backup archive.",
+				Title:   "Overwrite",
+				Example: "false",
+			},
+			plugin.Field{
+				Mode:    "target",
+				Name:    "prefix",
+				Type:    "string",
+				Help:    "This is the input string for prefix based backup-restore.",
+				Title:   "Prefix",
+				Example: "starkandwayne",
+			},
+		},
+	}
+	fmt.Fprintf(os.Stderr, "etcd plugin starting up...\n")
+	plugin.Run(p)
+}
+
+type EtcdPlugin plugin.PluginInfo
+
+type EtcdConfig struct {
+	EtcdClient      *clientv3.Client
+	TimeoutDuration time.Duration
+	Overwrite       bool
+	Prefix          string
+}
+
+func (p EtcdPlugin) Meta() plugin.PluginInfo {
+	return plugin.PluginInfo(p)
+}
+
+func getEtcdConfig(endpoint plugin.ShieldEndpoint) (*EtcdConfig, error) {
+	etcdEndpoint, err := endpoint.StringValue("url")
+	if err != nil {
+		return nil, err
+	}
+
+	etcdUrls := strings.Split(etcdEndpoint, ",")
+
+	timeoutStr, err := endpoint.StringValueDefault("timeout", "2")
+	if err != nil {
+		return nil, err
+	}
+
+	timeout, err := strconv.Atoi(timeoutStr)
+	if err != nil {
+		return nil, err
+	}
+
+	duration := time.Duration(timeout) * time.Second
+
+	auth, err := endpoint.StringValueDefault("auth", "")
+	if err != nil {
+		return nil, err
+	}
+
+	username, err := endpoint.StringValueDefault("username", "")
+	if err != nil {
+		return nil, err
+	}
+
+	password, err := endpoint.StringValueDefault("password", "")
+	if err != nil {
+		return nil, err
+	}
+
+	clientCert, err := endpoint.StringValueDefault("client_cert", "")
+	if err != nil {
+		return nil, err
+	}
+
+	clientKey, err := endpoint.StringValueDefault("client_key", "")
+	if err != nil {
+		return nil, err
+	}
+
+	caCert, err := endpoint.StringValueDefault("ca_cert", "")
+	if err != nil {
+		return nil, err
+	}
+
+	overwrite, err := endpoint.BooleanValueDefault("overwrite", false)
+	if err != nil {
+		return nil, err
+	}
+
+	prefix, err := endpoint.StringValueDefault("prefix", "")
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := &tls.Config{}
+	if auth == "Certificate-Based Authentication" {
+		caCertPool := x509.NewCertPool()
+		check := caCertPool.AppendCertsFromPEM([]byte(caCert))
+		if check != true {
+			plugin.DEBUG("CA cert did't parse right.\n")
+		}
```

Comment:
This should be set up if `ca_cert` is set to _anything_, in case I want to run etcd with my own internal CA-issued certificate, but do _not_ want to use anything other than role-based authentication (or don't auth at all!)